go/tools/gopackagesdriver/bazelquerydriver/driver.go

```diff
@@ -0,0 +1,323 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bazelquerydriver
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"go/types"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"github.com/bazelbuild/bazel-watcher/bazel"
+	"github.com/bazelbuild/bazel-watcher/third_party/bazel/master/src/main/protobuf/blaze_query"
+	"github.com/bazelbuild/rules_go/go/tools/gopackagesdriver/bazelquerydriver/pkgconv"
+	"github.com/bazelbuild/rules_go/go/tools/gopackagesdriver/protocol"
+	"golang.org/x/tools/go/packages"
+)
+
+const fileQueryPrefix = "file="
+
+var gorootPattern = regexp.MustCompile(`^bazel-[^/]+/external/go_sdk\b`)
+
+const supportedModes = packages.NeedName | packages.NeedFiles | packages.NeedCompiledGoFiles | packages.NeedImports | packages.NeedDeps | packages.NeedTypesSizes | packages.NeedModule
+
+type bazelDriver struct {
+	cfg           protocol.Request
+	resp          protocol.Response
+	bazel         bazel.Bazel
+	sdk           *gosdk
+	workspaceRoot string
+	bazelQueries  []string
+	stdlibImports map[string]bool
+	fileQueries   map[string]bool
+	importQueries map[string]bool
+	wildcardQuery bool
+}
+
+// New returns a Driver implementation based on bazel query.
+func New() protocol.Driver {
+	return func(cfg protocol.Request, patterns ...string) (*protocol.Response, error) {
+		bzl := bazel.New()
+		sdk, err := newGoSDK(bzl)
+		if err != nil {
+			return nil, err
+		}
+		driver := &bazelDriver{
+			cfg:           cfg,
+			bazel:         bzl,
+			sdk:           sdk,
+			stdlibImports: make(map[string]bool),
+			fileQueries:   make(map[string]bool),
+			importQueries: make(map[string]bool),
+		}
+		return driver.loadPackages(patterns...)
+	}
+}
+
+func (d *bazelDriver) loadPackages(patterns ...string) (*protocol.Response, error) {
+	if info, err := d.bazel.Info(); err != nil {
+		return &protocol.Response{NotHandled: true}, nil
+	} else {
+		d.workspaceRoot = info["workspace"]
+	}
+
+	log.Printf("mode: %v", d.cfg.Mode)
+	unsupportedModes := d.cfg.Mode &^ supportedModes
+	if unsupportedModes != 0 {
+		return nil, fmt.Errorf("%v (%b) not implemented", unsupportedModes, unsupportedModes)
+	}
+
+	if len(patterns) == 0 {
+		patterns = append(patterns, ".")
+	}
+
+	var ignoredPkg packages.Package
+
+	for _, patt := range patterns {
+		switch {
+		case strings.HasPrefix(patt, fileQueryPrefix):
+			fp := strings.TrimPrefix(patt, fileQueryPrefix)
+			if err := d.processFileQuery(fp, &ignoredPkg); err != nil {
+				return nil, err
+			}
+		case patt == ".":
+			d.bazelQueries = append(d.bazelQueries, goFilter(":all"))
+			d.wildcardQuery = true
+		case patt == "./...":
+			d.bazelQueries = append(d.bazelQueries, goFilter("..."))
+			d.wildcardQuery = true
+		case d.sdk.packages[patt]:
+			d.stdlibImports[patt] = true
+		default:
+			// TODO: do we need to check for a package ID instead of import path?
```

Comment:
I think we should assume arguments are Bazel labels, not import paths.

We'll need to handle import paths in the standard library, since there's no label for individual packages.

If we must process import paths without wildcards, we should make an attempt to find the directory without scanning the whole repo. `bazel query //...` can be very expensive for large repos.

Reply:
I'm definitely seeing some queries for import paths from gopls in my logfile.
How would we find the directory without scanning the whole repo?
Maybe we need to maintain a cache on disk of `import path -> package` and `generate filename -> package`?

Note that the first thing gopls does is query for "./..." which does take a long time. But then if you're lucky bazel keeps that loaded so later queries are fast.

Reply:
Hmm, I imagine we'll need to work with the gopls folks later on to avoid import path queries.

For now, let's add TODOs.

Reply:
I added a todo in sdk.go